v2/passwordless/common-customizations/sessions/user-impersonation.mdx

```diff
@@ -359,6 +359,109 @@ export default async function impersonate(req: SessionRequest, res: any) {
 }
 ```
 
+</TabItem>
+<TabItem value="nextjs-app">
+
+```tsx
+import { serialize } from "cookie";
+import { NextResponse, NextRequest } from "next/server";
+import SuperTokens from "supertokens-node";
+import { withSession } from "supertokens-node/nextjs";
+import { CollectingResponse, PreParsedRequest } from "supertokens-node/framework/custom";
+import { HTTPMethod } from "supertokens-node/types";
+import { createNewSession } from "supertokens-node/recipe/session";
+// @ts-ignore
+import { backendConfig } from "@/app/config/backend";
+
+SuperTokens.init(backendConfig());
+
+export function POST(req: NextRequest) {
+    return withSession(req, async (session) => {
+        const query = Object.fromEntries(new URL(req.url).searchParams.entries());
+        const cookies: Record<string, string> = Object.fromEntries(
+            req.cookies.getAll().map((cookie) => [cookie.name, cookie.value])
+        );
+
+        /**
+         * Pre parsed request is a wrapper exposed by SuperTokens. It is used as a helper to detect if the
+         * original request contains session tokens. We then use this pre parsed request to call `getSession`
+         * to check if there is a valid session.
+         */
+        let baseRequest = new PreParsedRequest({
+            method: req.method as HTTPMethod,
+            url: req.url,
+            query: query,
+            headers: req.headers,
+            cookies: cookies,
+            getFormBody: () => req!.formData(),
+            getJSONBody: () => req!.json(),
+        });
+
+        /**
+         * Collecting response is a wrapper exposed by SuperTokens. In this case we are using an empty
+         * CollectingResponse when calling `getSession`. If the request contains valid session tokens
+         * the SuperTokens SDK will attach all the relevant tokens to the collecting response object which
+         * we can then use to return those session tokens in the final result (refer to `withSession` in this file)
+         */
+        let baseResponse = new CollectingResponse();
+
+        let email = "..."; // read from request body
+
+        let users = await SuperTokens.listUsersByAccountInfo("public", { email });
+    
+        if (users.length === 0) {
+            throw new Error("User does not exist");
+        }
+
+        await createNewSession(baseRequest, baseResponse, "public", users[0].loginMethods[0].recipeUserId, {
```

Comment:
maybe we can add a createNewSession helper function also in our sdk to be something like this:
```
return await withCreateNewSession(req, (session) => {
   return NextResponse.json({  message: "Impersonation successful!" })
}, "public", userid, ...)  
```

- In this helper, we convert req => CollectingResponse like how we do here.
- The helper will create a CollectingResponse() object and then internally call createNewSession just like how it's done here.
- Then the helper will call the function provided by the user with the session object which then returns a nextjs object
- The helper will take the returned nextjs object and apply the headers and cookies to it as shown below.

So from a user's point of view, it will just be to call this helper function.

Reply:
Added `withPreParsedRequestResponse` helper function to achieve this instead.